### DISCUSSION TITLE
Slow upload speed through app

### DISCUSSION THREAD

**Discussion by jimski427** [immich-app/immich#8560] [Q&A]
### The bug

I have noticed quite a discrepancy with uploading files through the app with reverse proxy vs IP in the app and through the web. I thought it was due to my reverse proxy but we have ruled that out. when i upload through the web, i have tested the upload both inside and outside of the app. 
![Screenshot_20240405-222204](https://github.com/immich-app/immich/assets/51083907/0843c4a2-09c1-4a89-a4f8-1d4c2e6d7451)
![Screenshot_20240405-222114](https://github.com/immich-app/immich/assets/51083907/57604f4b-422b-40bb-a9b8-0c48e97b311d)

### The OS that Immich Server is running on

unraid

### Version of Immich Server

v1.101.0

### Version of Immich Mobile App

v1.101.0

### Platform with the issue

- [ ] Server
- [ ] Web
- [X] Mobile

### Your docker-compose.yml content

```YAML
ghcr.io/imagegenius/immich
```

### Your .env content

```Shell
ghcr.io/imagegenius/immich
```

### Reproduction steps

```bash
Use app with domain vs IP
```

**Comment by bo0tzz** [comment]
This seems unlikely to be a bug in Immich, so I've converted it to a discussion for now.

**Comment by jimski427** [comment]
If you could tell me what it could be that isn't immich that would be awesome but considering that the only time the upload is slow is through the app, with a domain, I would have to believe it's the app. I can upload via browser from the same device, to the same domain, and it uploads at 10x the speed.

**Reply by alextran1502** [reply]
We will have to know more about your network setup in order to help pin point where might be the problem

**Reply by jimski427** [reply]
We went through that on discord. I have pin pointed it to only the app, and only using my domain name.

**Reply by alextran1502** [reply]
If you upload through local IP and reach full speed, then have slower speed when using DNS, then the issue is somewhere between your DNS/proxy setup

**Reply by jimski427** [reply]
I reach full speed when uploading via DNS through the browser as well. It is ONLY uploading via domain in the app. The domain properly resolves to my local IP when ok my hole home network. Traceroutes to my domain show it going directly to my local IP with no hops.

**Reply by alextran1502** [reply]
Can you help us research the HTTP client of Dart to see if using reverse proxy/DNS can potentially cause this issue?

**Reply by jimski427** [reply]
I'm not familiar with that, but if you have any instructions or walk through, I can definitely work on it and provide whatever you would want.

**Reply by alextran1502** [reply]
Just research Google in general to see if the Dart HTTP client, the client we use for network communication in the mobile app, has some similar report issues with upload speed when using DNS/Reverse Proxy.

How did you confirm that uploading over the web got full speed? and what is your home network down/up speed?

**Reply by jimski427** [reply]
Will do. I confirmed it by uploading the same file both in the web and and via the app. I stopped the manual backup before it finished so I could test with the same file. Home ISP speeds are 1g down and 40mbps up, so if it was going out of my network, it would be limited to 40 Mbps or so, not 100-150. When I monitored the incoming traffic on my unRAID server, over the web it would hit around 800mbps. When I used the app, it maxes out at most, 150mbps. These are both going to the same exact domain name connected to my home WiFi network. Npm and immich are on the same machine, on the same custom docker network. Going to http://IP:port is functionally the same as https://domain in the browser. In the app, http://IP:port is the same as web, but when I use https://domain in the app, it slows down. Honestly, it wouldn't be a huge deal if manual backups were able to run in the background. But the fact that I have to babysit large manual file backups so the screen doesn't go dark or change apps, is a bit of a pain.

**Reply by jimski427** [reply]
Also this seems to be my exact issue. https://github.com/dart-lang/sdk/issues/54752

**Reply by mmomjian** [reply]
Sounds like this is an upstream issue, so there is not much we can do on the Immich side. It might help to add your observations there.

I would also recommend trying using HTTPS with a local IP, to see if that still slows down, as well as trying HTTP through NPM, to see what the speed of each one of these it and narrow it down to HTTPS traffic specifically, but it sounds like that might be too hard to do on your setup.

**Reply by LtDragon** [reply]
I use both immiach and photoprism on my truenas server.
For immiach I used the native app on my android and iPhone and my laptop app on wifi.
Vs
Photoprism with photosync on both and Andriod and iPhone and my laptop with the browser app on wifi.

I find uploading photos to photoprism is much faster and the server doesn't get fully utilized during the upload.
Vs 
Immiach the upload process is much longer and the server is 100% utilized during the upload process.

I would like to see a faster upload process immiach as I can see it's possible on a similar type software package. I don't mind the server being fully utilized but I'm not sure if it processing the images faster or it it just less efficient.

**Reply by Phlogi** [reply]
> Sounds like this is an upstream issue, so there is not much we can do on the Immich side. It might help to add your observations there.
> 
> I would also recommend trying using HTTPS with a local IP, to see if that still slows down, as well as trying HTTP through NPM, to see what the speed of each one of these it and narrow it down to HTTPS traffic specifically, but it sounds like that might be too hard to do on your setup.

You could switch to rhttp according https://github.com/dart-lang/sdk/issues/54752. 
Could immich do that? I can confirm that the upload is pretty slow.

**Reply by alextran1502** [reply]
The slow upload speed usually depends on your server/networking/disk setup. Below is an example of local upload

https://github.com/user-attachments/assets/9629f0c8-343f-45fd-a024-61d2eb92e4ec

**Reply by jimski427** [reply]
Imo that speed is abismal. I also have 1.5gb speed test between my phone and an open speed test container on the same unRAID server, the NIC on the unRAID server is 2.5gb and everything between the unRAID server and app is 2.5gb. I have tested it from the same device inside and outside the app and the issue only exists with the app. At this point I've just chalked it up to an app issue and don't really care about it since it's mostly uploading in the background anyways.

**Reply by Phlogi** [reply]
I cannot reproduce this either now when using that dialog. I think the background upload is slower, though, if that's even possible.

**Reply by schuettecarsten** [reply]
Any news on this issue?

**Comment by plopsadude** [comment]
same problem here, any sollution for this?

**Comment by terryhau** [comment]
I can confirm it's an issue with https.
I get 500 mbps over http and 50 mbps over https.

**Reply by back-lacking** [reply]
in a similar situation. 
going through my setup with https domain ans traefik as my rp, i got 5-10 mbps. 
going through straight http, i get 20-30 mbps

**Comment by mikael1022** [comment]
Yes same for me

**Comment by dasbootbee** [comment]
Same issue here.  Immich is behind a traefik reverse proxy using TLS.  Uploading from iOS on local Wi-Fi.

I just re-tested by exposing the ports in the docker-compose and connecting directly to the IP:port bypassing traefik using http.  The upload speed is the same.  Maxing out between 2-5 MB/s.

**Reply by datenzar** [reply]
I have the same setup (traefik as reverse proxy for ssl termination) and also observe a huge difference. 

In both cases, I access the immich server via local WiFi network. The ssl terminated route is maximum 11MB/sec whereas direct http via exposed port gives me about 50-80MB/sec.

Interesting to see that your ssl terminated speed is even half of mine. It might be due to the iPhone processing power. I have an iPhone 15 Pro, what do you have?

**Reply by dasbootbee** [reply]
I was using an iPhone 13 Pro.  I’m not sure that was the limiting factor but who knows?

**Comment by back-lacking** [comment]
has anyone found an workaround for this aside from using the http of server ip and port?
i would like to keep using https and my nice domain

**Comment by hugo-id** [comment]
Just to add to this,
I'm on Android, and have a seafile server on my server too.
It's all the same config (caddy, URL, https)
But where Seafile maxes my home internet connection (100Mbps), Immcih struggles to maintain 15Mbps

I managed to backup my entire 90GB library via seafile in maybe 5 hours (via LAN), where Immich hasn't finished syncing after _3 entire days!_

I don't understand how people could find this acceptable performance

**Reply by alextran1502** [reply]
I guess Seafile doesn't perform the following mechanism

- Generate thumbnails
- Generate encoded video
- Generate embedding information for CLIP search
- Facial detection
- Facial recognition
- OCR

So using Seafile doesn't maxing out CPU and I/O on your disk. 😉 

You can improve the process of uploading by

- Stoping the jobs (immich-go does this) to allow CPU and I/O doesn't get overwhelmed, and resume them after the upload is finished
- Using a CPU with more cores
- Using SSD as storage backend

Self-hosting server setup scenario is vast, before arriving to conclusion, I think it is best to share your server configuration for better understanding.

**Reply by dasbootbee** [reply]
Thanks for getting back to us on this.  If I understand what you're saying, all of these tasks happen before the next upload is initiated.  That would explain the different maximum speeds that all of us were seeing -- it's limited based on the back-end processing speed/storage.  Adding traefik/caddy + docker in the mix only adds more latency.

**Comment by AlexLJordan** [comment]
Hi!
I understand that Immich is hard at work to make all their cool features function.
But the original problem is that uploads are slow, and I'm still seeing this issue at least on my android phone.

So is there a way to tell Immich to take the upload first, before doing all it's post-processing?
Is there a way to configure a upload proxy that works at full speed, before handing files off to Immich, while still working properly through the app?

What I'm trying to say is; Knowing the cause is cool, but what do we do about it?
